Lib/contextlib.py

```diff
@@ -159,6 +165,12 @@ def __exit__(self, typ, value, traceback):
                 # tell if we get the same exception back
                 value = typ()
             try:
+                # If the generator handles the exception thrown into it, the
+                # exception context will revert to the actual current exception
+                # context here. In order to make the context manager behave
+                # like a normal function we set the current exception context
+                # to what it was during the context manager's __enter__
+                sys._set_exception(exc_context)
```

Comment:
TBH, I'm not that comfortable adding this function to the `sys` module, even with an `_`. It is an API that shouldn't really exist, we're adding it just to enable a hack in contextmanager, which needs to change the semantics of code.

Is there a way that we can roll this functionality into the `gen.throw()` function instead? Such as -- add a parameter that indicates that the exception stack needs to be fudged in this way?

Reply:
Yeah, that makes sense. I don't really have strong opinions on how it's implemented, I only went with `sys._set_exception` because it seemed like it would work and [there is precedent for a function that directly manipulates the current exception state]https://docs.python.org/2.7/library/sys.html#sys.exc_clear).

With that being said, I don't know if doing it entirely within `throw()` is possible, at least in this case. Because the exception that needs to be set is the one being handled in the `__enter__`, I don't think there's enough context when `throw()` is called to correctly manipulate the current exception.

Reply:
I was thinking you pass the exc_context as a new kwwarg to throw which defaults to None. 

(I wasn't aware of exc_clear actually. It was added in 2.3, so quite old. Doesn't seem to be used anywhere in the stdlib at the moment.)

Reply:
Ah, that makes sense. Now that I've got `sys._set_exception()` working and am therefore (reasonably) sure that the `None`/`NULL` exception stuff works properly, I'll take a look at adding a kwarg to `throw()` that essentially just makes it call `PyErr_SetHandledException` behind the scenes.